### PR TITLE
dev/core#252 - Add force search parameters in search forms

### PR DESCRIPTION
**Pull request by jitendrapurohit**
Overview
----------------------------------------
Add parameters to force search in Activity, Event, Membership & Contribution Search forms

Before
----------------------------------------
Some parameters were missing and cannot be specified in url param to search for the entities.

After
----------------------------------------
Parameters added and can be searched through URL params. Below is the table of field name vs params that can be used in the URL to force the search results.

**1. Contribution Tokens** -

|**Field Name**|**Param in URL**|
| ------------- |:-------------:| 
|    contribution_page_id | pid|
|    financial_type_id | fid|
|    contribution_status_id | status|
|    is_pay_later |pay_later|
|    receive_date_from (receive_date)|start|
|    receive_date_to (receive_date)|end|

Eg of complete search URL - 

    civicrm/contribute/search?reset=1&force=1&status=1&fid=2&pay_later=1&start=20181206&end=20181210

**2. Membership Tokens**

|**Field Name**|**Param in URL**|
| ------------- |:-------------:| 
|    membership_type_id | type |
|    membership_status_id |status|
|    start_date |start & start_high|
|    end_date |end & end_high|    

Eg of complete search URL - 

    civicrm/member/search?reset=1&force=1&type=1&status=1&start=20181002&start_high=20181004&end=20181103&end_high=20181109

**3. Participant Tokens**

|**Field Name**|**Param in URL**|
| ------------- |:-------------:| 
|    event_id | event |
|    status_id |status|
|    role_id |role|
|    is_pay_later |pay_later|    
|    register_date_from |register_start|    
|    register_date_to |register_end|    

Eg of complete search URL -

    civicrm/event/search?reset=1&force=1&event=1&status=2&role=3&pay_later=1&register_start=20181210&register_end=20181214

**4. Activity Tokens**

|**Field Name**|**Param in URL**|
| ------------- |:-------------:| 
|    activity-type id | type|
|    activity_status | status|
|    has a follow-up activity | hasFollowUp|
|    is a follow-up acivity |isFollowUp|
|    start date|dateLow|
|    end date|dateHigh|

Eg of complete search URL -

    civicrm/activity/search?reset=1&force=1&status=1&hasFollowUp=1&isFollowUp=1&dateLow=20181210&dateHigh=20181212&type=3

----------------

Gitlab Issue - https://lab.civicrm.org/dev/core/issues/252

### CODE REVIEW

File: CRM/Activity/BAO/Query.php [hunk 1]
```diff
@@ -335,11 +334,11 @@ public static function whereClauseSingle(&$values, &$query) {
 
       case 'parent_id':
         if ($value == 1) {
-          $query->_where[$grouping][] = "parent_id.parent_id IS NOT NULL";
+          $query->_where[$grouping][] = "civicrm_activity.parent_id IS NOT NULL";
           $query->_qill[$grouping][] = ts('Activities which have Followup Activities');
         }
         elseif ($value == 2) {
-          $query->_where[$grouping][] = "parent_id.parent_id IS NULL";
+          $query->_where[$grouping][] = "civicrm_activity.parent_id IS NULL";
```
Comment (jitendrapurohit) [inline CRM/Activity/BAO/Query.php]: This is a fix of a typo present in Activity search query. Leads to an error when a search is made on `is a Followup Activity?` field.

**Comment by eileenmcnaughton** [comment]
There is another search params issue on activity search (##11055) still hanging around because I'm not happy with us adding params to search without figuring out how to declare what parameters are acceptable for a form.

My thinking is we want something along the lines of #12185 that explicitly declares the fields to accept from a url and the validation to use for them

File: CRM/Contribute/Form/Search.php [hunk 2]
```diff
@@ -392,11 +392,16 @@ public function fixFormValues() {
     if (!$this->_force) {
       return;
     }
-
-    $status = CRM_Utils_Request::retrieve('status', 'String');
-    if ($status) {
-      $this->_formValues['contribution_status_id'] = array($status => 1);
-      $this->_defaults['contribution_status_id'] = array($status => 1);
+    $forceParams = [
+      'contribution_status_id' => ['name' => 'status', 'type' => 'String'],
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/Search.php]: We should use Alphanumeric instead of string

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit I agree with accepting more defaults but it is a security risk if not done robustly.

I proposed over here https://github.com/civicrm/civicrm-core/pull/12185/files a way to structure accepting input via a url (which in involved definining metadata on the form) and I would want to tie it in as an extension of the work done here https://github.com/civicrm/civicrm-core/pull/12078/files#diff-e0aa42922a99e97c7373924289bc0bf6R257

**Comment by jitendrapurohit** [comment]
@eileenmcnaughton updated the PR with `Alphanumeric` type.

I've tested #12185 which actually works well on the membership status page and also provides the security risk coverage. Though, on search forms, we have some elements which are not present in DB and still need a default to be set, eg `start_date_low`, `start_date_high`, date_relative input etc. Should this be handled separately as it was done before(or in this PR)?

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit per previous I've had a first cut at an approach like in #12185 for search forms https://github.com/civicrm/civicrm-core/pull/13013 (only one field so far). 
I think we can probably handle date ranges generically once we start looking at those fields

**Comment by eileenmcnaughton** [comment]
Oh - to your top table - hopefully we can get rid of some of the weird mappings. Down with 'pid' up with contribution_page_id :-)

**Comment by JoeMurray** [comment]
@colemanw would it be useful to develop a standard for mapping fields to url parameters? As a start, I'd suggest table_name.fieldname, where table_name is the civicrm table name minus the civicrm_ prefix. This will make the url long and ugly, but will avoid ambiguity and provide a path forward for a metadriven, general approach.

**Comment by petednz** [comment]
is this making progress again? can Fuzion help?

**Comment by eileenmcnaughton** [comment]
@petednz @jitendrapurohit so  I think #12185 creates the start of an approach & we should work on taking that further.

I guess the idea is that for each search form we'd define the metadata to be accessible from 

self::getSearchFieldMetadata() on the relevant query object. They are added to the form in buildSearchForm (on the same object)

```
    $form->addSearchFieldMetadata(['Grant' => self::getSearchFieldMetadata()]);
    $form->addFormFieldsFromMetadata();
    $form->convertFormValues();
```

This is where converted fields actually get set 

https://github.com/civicrm/civicrm-core/pull/13013/files#diff-4b584952e67c1c68b461b167faf234f3R415

**Comment by eileenmcnaughton** [comment]
I'm going to close this because it's gone stale - but I think the real reason it stalled is we hadn't agreed on an approach - I think we want the same metadata to be used to assign fields to a form as works to pick defaults as works to figure out what to render in a template - we've made some progress on that in ActivitySearch & grant search since this pr

Also we need tests if changing the Query object

**Pull request closed without merging**